boskos/aws-janitor/resources/launch_templates.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+)
+
+// LaunchTemplates https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DescribeLaunchTemplates
+type LaunchTemplates struct{}
+
+func (LaunchTemplates) MarkAndSweep(sess *session.Session, acct string, region string, set *Set) error {
+	svc := ec2.New(sess, &aws.Config{Region: aws.String(region)})
+
+	var toDelete []*launchTemplate // Paged call, defer deletion until we have the whole list.
+
+	pageFunc := func(page *ec2.DescribeLaunchTemplatesOutput, _ bool) bool {
+		for _, lt := range page.LaunchTemplates {
+			l := &launchTemplate{ID: *lt.LaunchTemplateId, Name: *lt.LaunchTemplateName}
+			if set.Mark(l) {
+				klog.Warningf("%s: deleting %T: %s", l.ARN(), lt, l.Name)
+				toDelete = append(toDelete, l)
+			}
+		}
+		return true
+	}
+
+	if err := svc.DescribeLaunchTemplatesPages(&ec2.DescribeLaunchTemplatesInput{}, pageFunc); err != nil {
+		return err
+	}
+
+	for _, lc := range toDelete {
+		deleteReq := &ec2.DeleteLaunchTemplateInput{
+			LaunchTemplateId:   aws.String(lc.ID),
+			LaunchTemplateName: aws.String(lc.Name),
+		}
+
+		if _, err := svc.DeleteLaunchTemplate(deleteReq); err != nil {
+			klog.Warningf("%s: delete failed: %v", lc.ARN(), err)
+		}
+	}
+
+	return nil
+}
+
+func (LaunchTemplates) ListAll(sess *session.Session, acct, region string) (*Set, error) {
+	c := ec2.New(sess, aws.NewConfig().WithRegion(region))
+	set := NewSet(0)
+	input := &ec2.DescribeLaunchTemplatesInput{}
+
+	err := c.DescribeLaunchTemplatesPages(input, func(lcs *ec2.DescribeLaunchTemplatesOutput, isLast bool) bool {
+		now := time.Now()
+		for _, lc := range lcs.LaunchTemplates {
+			arn := launchTemplate{
+				ID:   *lc.LaunchTemplateId,
+				Name: *lc.LaunchTemplateName,
+			}.ARN()
+			set.firstSeen[arn] = now
+		}
+
+		return true
+	})
+
+	return set, errors.Wrapf(err, "couldn't list launch templates for %q in %q", acct, region)
+}
+
+type launchTemplate struct {
+	ID   string
+	Name string
+}
+
+func (lc launchTemplate) ARN() string {
+	return lc.ID
```

Comment:
is this correct? does the ID itself already include the `arn:` prefix?

Reply:
I've updated it to include the full [launch template arn](https://docs.aws.amazon.com/IAM/latest/UserGuide/list_amazonec2.html#amazonec2-launch-template)